rust/src/cli/persist_nic.rs

```diff
@@ -195,49 +116,146 @@ pub(crate) fn run_persist_immediately(
     Ok("".to_string())
 }
 
-pub(crate) fn inspect(root: &str) -> Result<String, CliError> {
+fn gen_link_file_path(root: &str, iface_name: &str) -> PathBuf {
+    let link_dir = Path::new(root).join(SYSTEMD_NETWORK_LINK_FOLDER);
+
+    link_dir.join(format!("{PERSIST_FILE_PREFIX}-{iface_name}.link"))
+}
+
+fn extract_iface_names_from_link_file(file_name: &str) -> Option<String> {
+    file_name
+        .strip_prefix(&format!("{PERSIST_FILE_PREFIX}-"))
+        .and_then(|name| name.strip_suffix(".link"))
+        .map(ToOwned::to_owned)
+}
+
+pub(crate) fn clean_up(root: &str, dry_run: bool) -> Result<String, CliError> {
     let netdir = Path::new(root).join(SYSTEMD_NETWORK_LINK_FOLDER);
+
+    if !netdir.exists() {
+        log::info!("{} does not exist, no need to clean up", netdir.display());
+    }
     let stamp_path = netdir.join(NMSTATE_PERSIST_STAMP);
     if !stamp_path.exists() {
         log::info!(
-            "{} does not exist, no prior persisted state",
+            "{} does not exist, no prior persisted state, no need to clean up",
             stamp_path.display()
         );
         return Ok("".to_string());
     }
 
-    let mut n = 0;
+    let mut pinned_ifaces: HashMap<String, PathBuf> = HashMap::new();
+
     for e in netdir.read_dir()? {
         let e = e?;
-        let name = e.file_name();
-        let name = if let Some(n) = name.to_str() {
-            n
+        let file_name = if let Some(n) = e.file_name().to_str() {
+            n.to_string()
         } else {
             continue;
         };
-        if !name.ends_with(".link") {
-            continue;
-        }
-        if !name.starts_with(PERSIST_FILE_PREFIX) {
-            continue;
+        if let Some(iface_name) = extract_iface_names_from_link_file(&file_name)
+        {
+            log::info!("Found persisted NIC({iface_name}) file: {file_name}");
+            pinned_ifaces
+                .insert(iface_name.to_string(), netdir.join(file_name));
         }
-        log::info!("Found persisted NIC file: {name}");
-        n += 1;
     }
-    if n == 0 {
+
+    if pinned_ifaces.is_empty() {
         log::info!("No persisted NICs found");
+        return Ok("".to_string());
     }
 
-    let state = gather_state()?;
-    process_interfaces(&state, |iface, mac| {
-        let iface_name = iface.name();
-        log::info!(
-            "NOTE: would persist the interface with MAC {mac} to interface name {iface_name}"
-        );
-        Ok(())
+    for (iface_name, file_path) in pinned_ifaces.drain() {
```

Comment:
In most cases `into_iter()` is preferred over `drain` because it can be more efficient to also de-allocate the backing container at the same time.  It's actually the default for `for` loops https://doc.rust-lang.org/std/iter/index.html#for-loops-and-intoiterator so we can just drop the `.drain()` here.